File: labs/algorithms-live-coding/java/src/main/java/learning/livecoding/java/ExerciseSupport.java
package learning.livecoding.java;

public final class ExerciseSupport {
    private ExerciseSupport() {
    }

    public static void printHeading(Exercise exercise) {
        System.out.println("== " + exercise.id() + ": " + exercise.title() + " ==");
        System.out.println(exercise.summary());
    }

    public static void expectEquals(String label, Object expected, Object actual) {
        if (expected == null ? actual != null : !expected.equals(actual)) {
            throw new IllegalStateException(label + " expected <" + expected + "> but was <" + actual + ">");
        }
        System.out.println("PASS " + label + " -> " + actual);
    }

    public static void expectDouble(String label, double expected, double actual, double epsilon) {
        if (Math.abs(expected - actual) >= epsilon) {
            throw new IllegalStateException(label + " expected <" + expected + "> but was <" + actual + ">");
        }
        System.out.println("PASS " + label + " -> " + actual);
    }
}

File: labs/algorithms-live-coding/kotlin/src/main/kotlin/learning/livecoding/kotlin/ExerciseSupport.kt
package learning.livecoding.kotlin

object ExerciseSupport {
    fun printHeading(exercise: Exercise) {
        println("== ${exercise.id}: ${exercise.title} ==")
        println(exercise.summary)
    }

    fun expectEquals(label: String, expected: Any?, actual: Any?) {
        require(expected == actual) { "$label expected <$expected> but was <$actual>" }
        println("PASS $label -> $actual")
    }

    fun expectDouble(label: String, expected: Double, actual: Double, epsilon: Double = 0.000001) {
        require(kotlin.math.abs(expected - actual) < epsilon) {
            "$label expected <$expected> but was <$actual>"
        }
        println("PASS $label -> $actual")
    }
}
